### PR TITLE
feature: improve marketcap sorting using v3 cvs api

### PR DESCRIPTION
**Pull request by gre**
### 📝 Description

- This reimplement the `fetchMarketcapTickers` into `fetchIdsSortedByMarketcap`, moving from `/v2/tickers` to `/v3/currencies/supported` API which uses currency id instead of tickers. That itself should improve the marketcap ordering by using the true currency id. This shift created the opportunity to simplify the code of `sortByMarketcap`.
- Performance are **35 times faster** according to https://gist.github.com/gre/fe66ec88fa3de0767952591763bd0a67
```
currencies: 7915
v0_sortByMarketcap x 13.42 ops/sec ±0.58% (36 runs sampled)
v1_sortCurrenciesByIds x 5.61 ops/sec ±7.98% (19 runs sampled)
v2_sortCurrenciesByIds x 472 ops/sec ±3.76% (81 runs sampled)
```
- We also take the opportunity to remove some usage of `.countervalueEnabled` moving toward the idea that this boolean is entirely driven by the backend.
  - the only remaining usage of countervalueEnabled is related to crypto-icons (which have ticker collision still) so i'm not touching this yet.

- [ ] PR is currently blocked until we restore a "correct" ordering of the backend data. aka backend should return Bitcoin > Ethereum > ... and not Bitcoin > Arbitrum > ...

Screenshot 2023-11-14 at 16 29 07

### ❓ Context

- **JIRA or GitHub link**: https://ledgerhq.atlassian.net/browse/LIVE-9384 

### ✅ Checklist

Pull Requests must pass the CI and be code reviewed. Set as Draft if the PR is not ready.

- [x] `npx changeset` was attached.
- [x] **Covered by automatic tests.** 
- [x] **Impact of the changes:** 
  - LLD and LLM projects
  - all currencies list, typically in selects
  - the Manager app list

---

### 🧐 Checklist for the PR Reviewers

- [ ] **The code aligns with the requirements** described in the linked JIRA or GitHub issue.
- [ ] **The PR description clearly documents the changes** made and explains any technical trade-offs or design decisions.
- [ ] **There are no undocumented trade-offs**, technical debt, or maintainability issues.
- [ ] **The PR has been tested** thoroughly, and any potential edge cases have been considered and handled.
- [ ] **Any new dependencies** have been justified and documented.
- [ ] **Performance** considerations have been taken into account. (changes have been profiled or benchmarked if necessary)

### CODE REVIEW

**Review by KVNLS — CHANGES_REQUESTED** [review]
This algo is really tricky.
With current implementation it's slower than before.
image

https://github.com/LedgerHQ/ledger-live/assets/31533861/2d2f7608-1520-488f-8e89-264d5333d999


You can try to reproduce by clicking on Swap quick action

**Comment by gre** [comment]
@KVNLS makes sense. in future we need some benchmark tests.

i'll take a look on thurday. it will be easy to fix. at least the code is simpler but complexity probably went O(n2) on this part
```
  for (const id of ids) {
    const currency = currencies.find(c => c.type !== "FiatCurrency" && c.id === id);
    if (currency) sorted.push(currency);
  }
  ```
  
  can probably first map by id and then do the sorted push. moving to 2 loops of O(n) that way.

notice the fact we use `useCurrenciesByMarketcap` a bit everywhere nowadays (at the time we made this it was only done in top level if i recall) which means we do this multiple times (fetching from api AND sorting) maybe not the best idea overall but i'm not reconsidering this yet.

File: libs/ledger-live-common/src/currencies/sortByMarketcap.react.ts
Comment (haammar-ledger) [inline libs/ledger-live-common/src/currencies/sortByMarketcap.react.ts:1]: In some places we already use `.../hooks/...` for stuff that's specific to React, we should probably use only one pattern between this and `*.react.ts`

**Comment by gre** [comment]
I have reworked my code, indeed there were a performance regression (2.5x slower), my new implementation will be 35x faster 🚀 

proof https://gist.github.com/gre/fe66ec88fa3de0767952591763bd0a67

```
> node index.js

currencies: 7915
v0_sortByMarketcap x 13.42 ops/sec ±0.58% (36 runs sampled)
v1_sortCurrenciesByIds x 5.61 ops/sec ±7.98% (19 runs sampled)
v2_sortCurrenciesByIds x 472 ops/sec ±3.76% (81 runs sampled)
```

File: libs/ledger-live-common/src/currencies/sortByMarketcap.react.ts
Comment (gre) [reply libs/ledger-live-common/src/currencies/sortByMarketcap.react.ts:1]: for now i'm just splitting the file in two and keep things collocated, we can reconsider the file tree but it have bigger consequence than the scope here, i think we're just heading toward the idea that:

- step 1: split the react usage out into dedicated files
- step 2: reunify all into another library (or hooks/ files) , but best is actually to get rid of it all completely to make live-comon no longer react dependent

Screenshot 2023-11-16 at 11 26 19

File: libs/ledgerjs/packages/types-cryptoassets/src/index.ts [hunk 1]
```diff
@@ -224,7 +224,10 @@ type CurrencyCommon = {
   units: Unit[];
   // a shorter version of code using the symbol of the currency. like Ƀ . not all cryptocurrencies have a symbol
   symbol?: string;
-  // tells if countervalue need to be disabled (typically because colliding with other coins)
+  /**
+   * tells if countervalue need to be disabled (typically because colliding with other coins)
+   * @deprecated this field will soon be dropped. this is the API that drives this dynamically.
```
[the remark this answers is not in the document]
Comment (gre) [reply libs/ledgerjs/packages/types-cryptoassets/src/index.ts:229]: no issue.

**Review by KVNLS — APPROVED** [review]
Great improvements, congrats! This will improve a lot of user flows, this is even faster than before, thanks a lot!
(I only focused my review on sortByMarketCap and Swap flow)

File: libs/ledgerjs/packages/types-cryptoassets/src/index.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (gre) [reply libs/ledgerjs/packages/types-cryptoassets/src/index.ts:229]: maybe there is :D i can gen client side tho, weird

**Pull request merged**